### PR TITLE
Add AWS region to the AWS Config Cache key

### LINKED ISSUE #6128: AWS credentials cache key needs to include the region
### Report

Hello, I have two deployments using ScaledObjects based on SQS queues in different regions (say, eu-central-1 and us-east-1) and I want to authenticate to AWS using pod identity. The first ScaledObject authenticates correctly (one running on eu-central-1), and then the AWS config (with region included) is cached in the config cache. The second ScaledObject fails to start, because the operator tries to connect to a queue in another region (us-east-1), but the cached config includes the region from the first queue (eu-central-1). 
If I understand the code correctly, the `getCacheKey` function here: https://github.com/kedacore/keda/blob/85d4dca17f9e2e58bdc91f046e6dbe8e6235e78f/pkg/scalers/aws/aws_config_cache.go#L71 needs to include region in the returned string so that the configs are cached per region.

### Expected Behavior

Both ScaledObjects report as Ready

### Actual Behavior

Only the first ScaledObject is ready, the second one never authenticates successfully.

### Steps to Reproduce the Problem

1.Create two queues in different region
2.Create ScaledObjects for them using pod identity as auth mechanism

### Logs from KEDA operator

```
{"level":"error","ts":"2024-09-03T14:22:34Z","logger":"scale_handler","msg":"error getting metric for trigger","scaledObject.Namespace":"**REDACTED**","scaledObject.Name":"**REDACTED**","trigger":"awsSqsQueueScaler","error":"operation error SQS: GetQueueAttributes, https response error StatusCode: 400, RequestID: d511bcd7-6b16-5a65-bb0d-e38676fca9a3, AWS.SimpleQueueService.NonExistentQueue: The specified queue does not exist or you do not have access to it.","stacktrace":"github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).GetScaledObjectMetrics\n\t/workspace/pkg/scaling/scale_handler.go:553\ngithub.com/kedacore/keda/v2/pkg/metricsservice.(*GrpcServer).GetMetrics\n\t/workspace/pkg/metricsservice/server.go:48\ngithub.com/kedacore/keda/v2/pkg/metricsservice/api._MetricsService_GetMetrics_Handler.func1\n\t/workspace/pkg/metricsservice/api/metrics_grpc.pb.go:106\ngithub.com/kedacore/keda/v2/pkg/metricsservice.(*GrpcServer).Start.(*ServerMetrics).UnaryServerInterceptor.UnaryServerInterceptor.func6\n\t/workspace/vendor/github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/server.go:22\ngithub.com/kedacore/keda/v2/pkg/metricsservice/api._MetricsService_GetMetrics_Handler\n\t/workspace/pkg/metricsservice/api/metrics_grpc.pb.go:108\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/workspace/vendor/google.golang.org/grpc/server.go:1369\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/workspace/vendor/google.golang.org/grpc/server.go:1780\ngoogle.golang.org/grpc.(*Server).serveStreams.func2.1\n\t/workspace/vendor/google.golang.org/grpc/server.go:1019"}
```

### KEDA Version

2.15.0

### Kubernetes Version

1.30

### Platform

Amazon Web Services

### Scaler Details

AWS SQS

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by maxbog**
The PR adds the AWS region to the key used by the AWS config cache

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #6128

### CODE REVIEW

**Review by zroubalik — COMMENTED** [review]
The build is broken

**Comment by maxbog** [comment]
@zroubalik I fixed the build

**Comment by maxbog** [comment]
@zroubalik Is something wrong with the e2e tests? I see they failed twice now :(

**Comment by zroubalik** [comment]
> @zroubalik Is something wrong with the e2e tests? I see they failed twice now :(

yeah, issue on our side, will rerun once it is fixed

**Comment by ndlanier** [comment]
@zroubalik any updates on the e2e testing?

**Review by JorTurFer — APPROVED** [review]

**Comment by maxbog** [comment]
@JorTurFer I rebased the PR
Now the tests failed, but it looks like some random crash, can you please rerun the failing job and the e2e tests?

**Review by JorTurFer — APPROVED** [review]
PTAL @kedacore/keda-core-contributors

**Comment by ndlanier** [comment]
Any insight into why it panic'd/failed on the amd64 validation?

**Comment by ndlanier** [comment]
@JorTurFer is the validate - amd64 stage failure an issue with the action or an issue with the code changes in this pr?

**Comment by JorTurFer** [comment]
> @JorTurFer is the validate - amd64 stage failure an issue with the action or an issue with the code changes in this pr?

I've triggered it again because probably it was a transient error

**Comment by JorTurFer** [comment]
@zroubalik @wozniakjan PTAL

**Comment by ThaSami** [comment]
i would suggest to add Region in aws authorization metadata, rather than passing it all around. 

https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws/aws_authorization.go#L18C1-L35

and init it here 
https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws/aws_common.go#L53C1-L59C1

wdyt @JorTurFer ?

**Comment by JorTurFer** [comment]
> i would suggest to add Region in aws authorization metadata, rather than passing it all around.
> 
> [`5ea1eac`/pkg/scalers/aws/aws_authorization.go#L18C1-L35](https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws/aws_authorization.go#L18C1-L35)
> 
> wdyt @JorTurFer ?

That's a good idea tbh. Could you update the PR @maxbog ?

**Comment by maxbog** [comment]
@JorTurFer @ThaSami 
Thanks for the review. 
At first glance, I also thought this might be a good idea. However, after digging into the code, I'm not that convinced anymore. Here are the reasons:
1. Although the `AuthorizationMetadata` struct has `Authorization` in the name, what it really represents is an AWS identity: a role + credentials coming from pod identity or the AccessKey/SecretAccessKey pair. Identities are global in AWS, so IMO, region does not belong in this struct. Region is actually only used when creating AWS clients, to specify, where the queried services are, not where the identity is located.
2. We would get rid of the `awsRegion` parameter in the `getCacheKey` function, but then, if we still want to add it to the `AuthorizationMetadata`, we need to add a `awsRegion` parameter to the [`GetAwsAuthorization`](https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws/aws_common.go#L86) function, which, in turn, means we have to pass it from the different types of scaler metadata (for example, [here](https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws_dynamodb_scaler.go#L113), [here](https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/kafka_scaler.go#L335), or [here](https://github.com/kedacore/keda/blob/5ea1eac06e536a04cdf22383faff9e91bd374f17/pkg/scalers/aws_sqs_queue_scaler.go#L181), or 5 different places)
3. Having the `awsRegion` parameter in the `getCacheKey` actually makes sense when you look at the `AuthorizationMetadata` as an identity. The `aws.Config` struct represents a AWS service client configuration, so IMO it makes perfect sense to cache it for a specific identity and for a specific region. We lose expressiveness by hiding the region in the `awsAuthorization` parameter.

Please let me know, if, after taking the above into account, you still want to update the PR and add the region to the `AuthorizationMetadata`.

**Comment by ThaSami** [comment]
> Although the AuthorizationMetadata struct has Authorization in the name, what it really represents is an AWS identity: a role + credentials coming from pod identity or the AccessKey/SecretAccessKey pair.

Exactly - that's what sparked this idea for me. Since credentials are part of identity, obtaining them in the AWS SDK requires specifying a region as STS endpoints are regional. 
that's why i believe including AwsRegion within the AuthorizationMetadata struct makes architectural sense because:
  1. keeps all authentication-related data in one cohesive unit
  2. Ensures the region information is always available when making credential requests
  3. maintains a clear relationship between authorization data and its associated AWS region
  
 would like to hear from @JorTurFer

**Review by zroubalik — COMMENTED** [review]
I like the proposal from @ThaSami

**Comment by JorTurFer** [comment]
> I like the proposal from @ThaSami

Me too :)

**Comment by maxbog** [comment]
Ok then, I'll make the change

**Comment by maxbog** [comment]
One more question: 
Now that the region will be embedded in the `AuthorizationMetadata`, I think I should get rid of `awsConfigMetadata`, as it's now redundant. Do you agree?
https://github.com/kedacore/keda/blob/68b6ca8d70950cb337312b5efdb7056c861af9fb/pkg/scalers/aws/aws_common.go#L42-L45

**Comment by JorTurFer** [comment]
> Do you agree?

yeah, makes sense!

**Comment by maxbog** [comment]
@JorTurFer @ThaSami please take another look

File: pkg/scalers/aws/aws_common.go
Comment (JorTurFer) [inline pkg/scalers/aws/aws_common.go:49]: The empty line is detected by static checks
```suggestion
func GetAwsConfig(ctx context.Context, awsAuthorization AuthorizationMetadata) (*aws.Config, error) {
```

File: CHANGELOG.md
Comment (JorTurFer) [inline CHANGELOG.md:101]:
```suggestion
- **General**: Scalers cache uses a mutex to prevent concurrent actions ([#6273](https://github.com/kedacore/keda/issues/6273))
- **AWS Scalers**: Add AWS region to the AWS Config Cache key ([#6128](https://github.com/kedacore/keda/issues/6128))
- **AWS CloudWatch Scaler**: Properly parse multiple values of `dimensionName` & `dimensionValue` lists ([#6236](https://github.com/kedacore/keda/issues/6236))
```

**Comment by JorTurFer** [comment]
Unit test are failing, could you take a look?

File: CHANGELOG.md
Comment (maxbog) [reply CHANGELOG.md:101]: I'm not 100% sure, but won't that suggestion fail the CI build?

**Comment by maxbog** [comment]
> Unit test are failing, could you take a look?

The failing tests are for prometheus scaler, and the error message is: "HTTP transport must be Google OAuth2"
TBH, I'm not sure how this is related to the changes in the PR.

**Comment by JorTurFer** [comment]
Prometheus scaler uses multiple auth providers, maybe the region change of the latest commit has had a side effect here: https://github.com/kedacore/keda/blob/68b6ca8d70950cb337312b5efdb7056c861af9fb/pkg/scalers/prometheus_scaler.go#L96-L128

**Comment by JorTurFer** [comment]
Most probably it's returning an error here:
https://github.com/kedacore/keda/blob/68b6ca8d70950cb337312b5efdb7056c861af9fb/pkg/scalers/prometheus_scaler.go#L119-L123
An error in that line will make fail the test although is GCP because it returns error and not the GCP transport

**Comment by maxbog** [comment]
Unit tests are passing now

File: pkg/scalers/aws/aws_common.go
Comment (maxbog) [reply pkg/scalers/aws/aws_common.go]: Dixed

File: CHANGELOG.md
Comment (JorTurFer) [reply CHANGELOG.md:101]: f**k! yes, you're right!

Comment (JorTurFer) [reply CHANGELOG.md:101]: It should be alphabetically sorted. My point was to move the current like from **General** to *AWS Scalers*

Comment (maxbog) [reply CHANGELOG.md]: ah, ok, I changed the scaler name

File: CHANGELOG.md
Comment (zroubalik) [inline CHANGELOG.md:101]: 2.16 is already released, we should move this to unreleased

Comment (maxbog) [reply CHANGELOG.md]: done

**Comment by maxbog** [comment]
Can we merge the PR? I'd rather not be stuck in the conflict -> merge -> e2e test -> conflict death spiral ;)

**Comment by votzest** [comment]
it seems e2e tests got stuck again? otherwise can you merge this pr @zroubalik, please?

**Comment by JorTurFer** [comment]
@zroubalik @wozniakjan , PTAL

**Comment by maxbog** [comment]
@JorTurFer @zroubalik Please take a look at this PR. It's waiting for approval for 3 weeks now. The change is ready, waiting to be merged.

**Review by zroubalik — APPROVED** [review]

**Comment by zroubalik** [comment]
@maxbog there are some conflicts, please fix them and then I think we can merge this.

**Comment by maxbog** [comment]
@zroubalik done, no conflicts now

**Comment by zroubalik** [comment]
No, we don't include a new features in the patch releases. This will be included in 2.17.

**Comment by ndlanier** [comment]
I would argue this is a fix not a feature given that it's being implemented to account for a bug in handling cached credentials when the next target in an aws environment is in a different region.
https://github.com/kedacore/keda/issues/6128

Are there any ETA's or target dates for 2.17?

**Pull request merged**
